pkg/provider/cloud/vsphere/tags.go

```diff
@@ -29,78 +29,158 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-func reconcileTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
-	if err := syncCreatedClusterTags(ctx, restSession, cluster); err != nil {
-		return nil, fmt.Errorf("failed to sync created tags %w", err)
+func reconcileTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, sets.Set[string], error) {
+	var err error
+	managedTags, err := syncCreatedClusterTags(ctx, restSession, cluster)
+	if err != nil {
+		return cluster, nil, fmt.Errorf("failed to sync created tags: %w", err)
 	}
 
-	if err := syncDeletedClusterTags(ctx, restSession, cluster); err != nil {
-		return nil, fmt.Errorf("failed to sync deleted tags %w", err)
+	managedTags, err = syncDeletedClusterTags(ctx, restSession, cluster, managedTags)
+	if err != nil {
+		return cluster, nil, fmt.Errorf("failed to sync deleted tags: %w", err)
 	}
 
-	cluster, err := update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
+	// Avoid nil pointer panic
+	var managedTagIDs []string
+	if cluster.Status.ProviderStatus == nil || cluster.Status.ProviderStatus.VSphere == nil || cluster.Status.ProviderStatus.VSphere.ManagedTagIDs == nil {
+		managedTagIDs = []string{}
+	} else {
+		managedTagIDs = cluster.Status.ProviderStatus.VSphere.ManagedTagIDs
+	}
+	currentTagsSet := sets.New(managedTagIDs...)
+	if !currentTagsSet.Equal(managedTags) {
+		if cluster.Status.ProviderStatus == nil {
+			cluster.Status.ProviderStatus = &kubermaticv1.ClusterProviderStatus{
+				VSphere: &kubermaticv1.VSphereStatus{},
+			}
+		}
+		if cluster.Status.ProviderStatus.VSphere == nil {
+			cluster.Status.ProviderStatus.VSphere = &kubermaticv1.VSphereStatus{}
+		}
+		cluster.Status.ProviderStatus.VSphere.ManagedTagIDs = sets.List(managedTags)
+		return cluster, managedTags, nil
+	}
+	updatedCluster, err := update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
 		if !kuberneteshelper.HasFinalizer(cluster, tagCleanupFinalizer) {
 			kuberneteshelper.AddFinalizer(cluster, tagCleanupFinalizer)
 		}
 	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to add finalizer %s on vsphere cluster object: %w", tagCleanupFinalizer, err)
-	}
-
-	return cluster, nil
+	return updatedCluster, managedTags, err
```

Comment:
> cluster.Status.ProviderStatus.VSphere.ManagedTagIDs = sets.List(managedTags)

Since `cluster.Status.ProviderStatus.VSphere.ManagedTagIDs` is updated with `sets.List(managedTags)`, why return `managedTags` separately? The caller can access the updated tags directly from the cluster status, making the return parameter redundant.